ppy/osu
# Implement multiplayer countdown timers

- Resolves https://github.com/ppy/osu/issues/16408
- Requires osu-server-spectator deploy with https://github.com/ppy/osu-server-spectator/pull/109.

Preview: https://user-images.githubusercontent.com/1329837/158999082-07273668-fc10-4709-887b-7aac1c00ac78.mp4

Provides the basic structure for countdown timers, and support for a countdown timer to automatically start the match.

- `MultiplayerRoom` has a new `Countdown` member which stores basic information about the countdown, such as when it ends.
    - Of abstract type `MultiplayerCountdown`.
    - I expect this type to be derived for other purposes in the future, such as if we want to support a description in more "simple" countdown types.
- `MatchStartCountdown` derives from `MultiplayerCountdown` and indicates a match start countdown to the client.
- Changes to the countdown (starting or finishing) are communicated to client via `CountdownChangedEvent`.
- The host can trigger a countdown via `StartMatchCountdownRequest` and cancel it via `CancelCountdownRequest`.

Designs are pending, however for the time being I've split the ready button into two once the host readies up to show a dropdown button with options to start a countdown lasting for 10sec, 30sec, 1min, or 2min. Once a countdown is initiated, clicking the ready button again cancels the countdown.

Full testing procedure (client + server):
```
Solo:
- Start match normally.
    - Should start.
- Start match with 10s countdown.
    - Should start.
- Cancel 10s countdown.
    - Should cancel countdown.
- Start countdown and change to spectating.
    - Should cancel countdown.
- 2 minute countdown.
    - Should start.

Multi:
- Start 10s countdown, join other user.
    - Should show ongoing countdown for other user, not start match for other user.
- Start 10s countdown, join other user, ready other user.
    - Should start match for other user.
- Start 10s countdown, ready other user.
    - Should start match for other user.
- Start 10s countdown, ready other user, leave other user.
    - Should start match for the host.
- Start 10s countdown, ready then unready other user.
    - Should not start match for the other user.
- Start 10s countdown, change other user to spectating.
    - Both users should start match.
- Start 10s countdown, ready other user, change host to spectating.
    - Both users should start match.
- Start 10s countdown, ready other user, change host to spectating, then remove spectating from host.
    - Should cancel countdown.
```
